openhtf/core/phase_executor.py

```diff
@@ -185,47 +191,52 @@ def __init__(self, test_state):
     self._stopping = threading.Event()
 
   def execute_phase(self, phase):
-    """Executes each phase or skips them, yielding PhaseOutcome instances.
+    """Executes a phase or skips it, yielding PhaseExecutionOutcome instances.
 
     Args:
       phase: Phase to execute.
 
     Returns:
-      The final PhaseOutcome that wraps the phase return value (or exception)
-      of the final phase run. All intermediary results, if any, are REPEAT
-      and handled internally. Returning REPEAT here means the phase hit its
-      limit for repetitions.
+      The final PhaseExecutionOutcome that wraps the phase return value
+      (or exception) of the final phase run. All intermediary results, if any,
+      are REPEAT and handled internally. Returning REPEAT here means the phase
+      hit its limit for repetitions.
     """
     repeat_count = 1
     repeat_limit = phase.options.repeat_limit or sys.maxint
     while not self._stopping.is_set():
-      outcome = self._execute_phase_once(phase)
+      is_last_repeat = repeat_count >= repeat_limit
+      phase_execution_outcome = self._execute_phase_once(phase, is_last_repeat)
 
-      if outcome.is_repeat and repeat_count < repeat_limit:
+      if phase_execution_outcome.is_repeat and not is_last_repeat:
         repeat_count += 1
         continue
 
-      return outcome
+      return phase_execution_outcome
     # We've been cancelled, so just 'timeout' the phase.
-    return PhaseOutcome(None)
+    return PhaseExecutionOutcome(None)
 
-  def _execute_phase_once(self, phase_desc):
-    """Executes the given phase, returning a PhaseOutcome."""
+  def _execute_phase_once(self, phase_desc, is_last_repeat):
+    """Executes the given phase, returning a PhaseExecutionOutcome."""
     # Check this before we create a PhaseState and PhaseRecord.
     if phase_desc.options.run_if and not phase_desc.options.run_if():
       _LOG.info('Phase %s skipped due to run_if returning falsey.',
                 phase_desc.name)
-      return PhaseOutcome(openhtf.PhaseResult.CONTINUE)
+      return PhaseExecutionOutcome(openhtf.PhaseResult.SKIP)
 
     with self.test_state.running_phase_context(phase_desc) as phase_state:
       _LOG.info('Executing phase %s', phase_desc.name)
       phase_thread = PhaseExecutorThread(phase_desc, self.test_state)
       phase_thread.start()
       self._current_phase_thread = phase_thread
-      phase_state.result = phase_thread.join_or_die()
-
-    _LOG.debug('Phase finished with result %s', phase_state.result)
-    return phase_state.result
+      result = phase_state.result = phase_thread.join_or_die()
+      if phase_state.result.is_repeat and is_last_repeat:
```

Comment:
this feels like it should be in `execute_phases`, not `execute_phase_once`, mostly because this function doesn't know or care about repeats outside of this small snippet, while the other is mostly written around repeats